File: eval-platform/src/main/java/com/sap/hackaton2025/player/service/ApiService.java
package com.sap.hackaton2025.player.service;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sap.hackaton2025.player.model.RoundRequest;
import com.sap.hackaton2025.player.model.RoundResponse;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class ApiService {
    private static final String API_KEY = "ee3dd8b7-9e63-4054-b976-cfa35cafb3c2"; 
    private static final String BASE_URL = "http://127.0.0.1:8080/api/v1";
    
    private final HttpClient client = HttpClient.newHttpClient();
    private final ObjectMapper mapper = new ObjectMapper();
    private String sessionId = "";

    public void startSession() throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/session/start"))
                .header("API-KEY", API_KEY)
                .POST(HttpRequest.BodyPublishers.noBody())
                .build();

        HttpResponse<String> resp = client.send(request, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 200) {
            String body = resp.body().trim();
            if (body.startsWith("{")) {
                JsonNode node = mapper.readTree(body);
                if (node.has("sessionId")) sessionId = node.get("sessionId").asText();
                else if (node.has("id")) sessionId = node.get("id").asText();
                else sessionId = body;
            } else {
                sessionId = body.replace("\"", "");
            }
        } else {
            throw new RuntimeException("Start Failed: " + resp.body());
        }
    }

    public void endSession() {
        try {
            if (sessionId.isEmpty()) return;
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(BASE_URL + "/session/end"))
                    .header("API-KEY", API_KEY)
                    .header("SESSION-ID", sessionId)
                    .POST(HttpRequest.BodyPublishers.noBody())
                    .build();
            client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            System.err.println("Error closing session: " + e.getMessage());
        }
    }

    public RoundResponse playRound(RoundRequest payload) throws Exception {
        String jsonBody = mapper.writeValueAsString(payload);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/play/round"))
                .header("API-KEY", API_KEY)
                .header("SESSION-ID", sessionId)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();

        HttpResponse<String> resp = client.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (resp.statusCode() == 400 && resp.body().contains("Session already ended")) {
            return null; 
        }
        if (resp.statusCode() != 200) {
            throw new RuntimeException("Server Error " + resp.statusCode() + ": " + resp.body());
        }
        return mapper.readValue(resp.body(), RoundResponse.class);
    }
    
    public String getSessionId() { return sessionId; }
}

File: eval-platform/python_client/player/services/api_service.py
import requests
import json
from player.models import RoundResponse

class ApiService:
    API_KEY = "101f9a4d-242e-4d4a-83c5-409d58153701"
    BASE_URL = "http://127.0.0.1:8080/api/v1"
    
    def __init__(self):
        self.session_id = ""
        self.headers = {
            "API-KEY": self.API_KEY,
            "Content-Type": "application/json"
        }

    def start_session(self):
        url = f"{self.BASE_URL}/session/start"
        # POST with empty body
        resp = requests.post(url, headers=self.headers)
        
        if resp.status_code == 200:
            body = resp.text.strip()
            # Handle potential JSON response or plain string
            if body.startswith("{"):
                data = resp.json()
                self.session_id = data.get("sessionId") or data.get("id") or body
            else:
                self.session_id = body.replace('"', '')
            
            # Update headers with session ID for future requests
            self.headers["SESSION-ID"] = self.session_id
        else:
            raise Exception(f"Start Failed: {resp.text}")

    def end_session(self):
        if not self.session_id: return
        url = f"{self.BASE_URL}/session/end"
        try:
            requests.post(url, headers=self.headers)
        except Exception as e:
            print(f"Error ending session: {e}")

    def play_round(self, round_request):
        url = f"{self.BASE_URL}/play/round"
        
        payload = round_request.to_dict()
        
        # Trimitem cererea
        resp = requests.post(url, headers=self.headers, json=payload)
        
        # Verificăm dacă jocul s-a terminat
        if resp.status_code == 400 and "Session already ended" in resp.text:
            return None, None # <-- Returnăm dublu None
            
        if resp.status_code != 200:
            raise Exception(f"Server Error {resp.status_code}: {resp.text}")
            
        data = resp.json() 
        
        # Returnăm TUPLU: (Obiectul procesat, Datele brute)
        return RoundResponse(data), data

    def get_session_id(self):
        return self.session_id